### PR TITLE
feat: Remove `comptime` and warn upon usage

### LINKED ISSUE #280: Track IsConst for `==` and `!=` on arrays
Currently if the user writes `array1 == array2` or `array1 != array2` (given both variables are arrays), the result will never be const, even if both arrays have const elements.

In practice, this is only noticeable if the result of this expression is later casted to an integer and used as an array index:
```
let i = (array1 == array2) as Field;
constrain array1[i] == foo; // error: i is not const
```

### LINKED ISSUE #793: Comptime soundness issue in `if` expressions
# Description

## Aim

```rs
fn main(x: bool) {
    let foo = if x { 1 } else { 2 };
    ...
}
```

## Expected behavior

Expected `foo` to be a `Field` rather than a `comptime Field` since its value depends on a private input to the circuit.

## Bug

`foo` is incorrectly typed as a `comptime Field` which can cause compiler panics when later attempting to use it as an array index.

### LINKED ISSUE #1035: Unreachable code: logging this expression type is not supported
### Aim

Trying to compile the program:
```rs
use dep::std;

struct RV<T> {
    value: T,
    count: comptime Field,
}

impl<T> RV<T> {
    fn new(value: T) -> Self {
        Self { value, count: 1 }
    }

    fn increment(mut this: Self) -> Self {
        this.count += 1;
        this
    }

    fn print(self) {
        for _i in 0 .. self.count {
            std::println(self.value);
        }
    }
}

fn main() {
    let mut rv = RV::new(5);
    rv = rv.increment();
    rv.print();
}
```

### Expected behavior

Expected the program to be compile and for 5 to be printed twice.

### Bug

Compiler crashes after hitting an `unreachable!` line:
```
The application panicked (crashed).
Message:  internal error: entered unreachable code: logging this expression type is not supported
Location: crates/noirc_evaluator/src/ssa/context.rs:1165

This is a bug. Consider opening an issue at https://github.com/noir-lang/noir/issues/new?labels=bug&template=bug_report.yml

  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ BACKTRACE ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
                                ⋮ 13 frames hidden ⋮                              
  14: noirc_evaluator::ssa::context::SsaContext::get_builtin_opcode::h81d6c1bcd20b2110
      at /Users/jakefecher/Code/Noir/noir/crates/noirc_evaluator/src/ssa/context.rs:1165
  15: noirc_evaluator::ssa::function::<impl noirc_evaluator::ssa::ssa_gen::IrGenerator>::call::h1b26b3b2fd8da2c5
      at /Users/jakefecher/Code/Noir/noir/crates/noirc_evaluator/src/ssa/function.rs:233
  16: noirc_evaluator::ssa::ssa_gen::IrGenerator::ssa_gen_expression::h27c0e2dc747eea02
      at /Users/jakefecher/Code/Noir/noir/crates/noirc_evaluator/src/ssa/ssa_gen.rs:551
  17: noirc_evaluator::ssa::ssa_gen::IrGenerator::ssa_gen_expression::h27c0e2dc747eea02
      at /Users/jakefecher/Code/Noir/noir/crates/noirc_evaluator/src/ssa/ssa_gen.rs:575
  18: noirc_evaluator::ssa::ssa_gen::IrGenerator::ssa_gen_block::he32ad37f040a9f9e
      at /Users/jakefecher/Code/Noir/noir/crates/noirc_evaluator/src/ssa/ssa_gen.rs:737
  19: noirc_evaluator::ssa::ssa_gen::IrGenerator::ssa_gen_expression::h27c0e2dc747eea02
      at /Users/jakefecher/Code/Noir/noir/crates/noirc_evaluator/src/ssa/ssa_gen.rs:562
  20: noirc_evaluator::ssa::ssa_gen::IrGenerator::ssa_gen_for::h20d6e02271f201c2
      at /Users/jakefecher/Code/Noir/noir/crates/noirc_evaluator/src/ssa/ssa_gen.rs:702
  21: noirc_evaluator::ssa::ssa_gen::IrGenerator::ssa_gen_expression::h27c0e2dc747eea02
      at /Users/jakefecher/Code/Noir/noir/crates/noirc_evaluator/src/ssa/ssa_gen.rs:554
  22: noirc_evaluator::ssa::ssa_gen::IrGenerator::ssa_gen_block::he32ad37f040a9f9e
      at /Users/jakefecher/Code/Noir/noir/crates/noirc_evaluator/src/ssa/ssa_gen.rs:737
  23: noirc_evaluator::ssa::ssa_gen::IrGenerator::ssa_gen_expression::h27c0e2dc747eea02
      at /Users/jakefecher/Code/Noir/noir/crates/noirc_evaluator/src/ssa/ssa_gen.rs:562
```

### To reproduce

1.
2.
3.
4.

### Installation method

None

### Nargo version

_No response_

### @noir-lang/noir_wasm version

_No response_

### @noir-lang/barretenberg version

_No response_

### @noir-lang/aztec_backend version

_No response_

### Additional context

_No response_

### Submission Checklist

- [ ] Once I hit submit, I will assign this issue to the Project Board with the appropriate tags.

### LINKED ISSUE #1188: Cannot assign an expression of type Field to a value of type comptime Field
### Aim

Run `nargo check` on the following contrived example:

```rust
fn f(i: Field, j: comptime Field)
{
    let mut x = 0;

    let _pred = (x == j);

    x = i;

}
```

### Expected behavior

Constraint system should be successfully built.

### Bug

We get the following error pointing to the line `x = i;`:
`error: Cannot assign an expression of type Field to a value of type comptime Field`

### To reproduce

1. Place the above function in a Noir project.
2. Run `nargo check`.

### Installation method

Compiled from source

### Nargo version

nargo 0.4.0 (git version hash: 528a2a441cfe094885cc8f26ffba865f3a0b5c0c, is dirty: false)

### @noir-lang/noir_wasm version

_No response_

### @noir-lang/barretenberg version

_No response_

### @noir-lang/aztec_backend version

_No response_

### Additional context

The error goes away if a type annotation is provided for `x`:

```rust
let mut x: Field = 0;
```

Doing this leads to unexpected values being assigned to `x` in more elaborate examples, but I do not have a minimal example illustrating this at present. Also note that the error goes away if the line involving the comparison (`let _pred = (x == j);`) is commented out.

### Submission Checklist

- [ ] Once I hit submit, I will assign this issue to the Project Board with the appropriate tags.

### LINKED ISSUE #2040: Remove comptime feature
### Problem

This is a tracking issue to remove comptime from the language.

### Happy Case

.

### Alternatives Considered

_No response_

### Additional Context

_No response_

### Would you like to submit a PR for this Issue?

No

### Support Needs

_No response_

### PR DESCRIPTION
**Pull request by jfecher**
# Description

## Problem

Resolves #2040 
Resolves #280
Resolves #1188
Resolves #793
Resolves #1035 - the failure to unroll the for loop is now an error instead of a panic

## Summary

This PR removes `comptime` from the language. The `comptime` keyword and syntax are currently still kept and parsed for backwards compatibility, but are now deprecated and will issue a warning when used. `comptime` has been removed because it is no longer needed for accessing arrays. However, since for loops still must be known at compile-time I've taken the liberty of adding an error message for this which also solves some past panics that the `comptime` tracking did not catch.

Example error message for a loop whose length is not known at compile-time:

```
error: 
   ┌─ /.../noir/short/src/main.nr:28:22
   │
28 │         for _i in 0..self.count {
   │                      ---------- Could not determine loop bound at compile-time
```
The error message will always correctly point to only the part of the range that is unknown (start or end). If both are unknown, only the start range error is issued since we halt on the first error.

## Documentation

- [x] This PR requires documentation updates when merged.

  

  - [ ] I will submit a noir-lang/docs PR.

  

  - [x] I will request for and support Dev Rel's help in documenting this PR.
    - All references to `comptime` can be removed. We should still mention that the bounds of a for loop need to be known at compile-time. Since the compiler inlines every function, function parameters are still okay for use as for loop bounds.
    - We should also remove the `RepeatedValue` example. It has been broken in Noir for a while.
 
  
  

## Additional Context

I'd like some discussion on this PR. Removing `comptime` simplifies the system and removes the confusing subtle differences between `comptime` and rust's `const`, but also means it may be more difficult for users to track why a for loop's bounds may or may not be known. Instead of being able to follow `comptime` annotations in the program, they will only get an error on the for loop itself and would have to work backward to figure out why the bounds are not compile-time known.

# PR Checklist\*

- [x] I have tested the changes locally.
- [x] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

**Comment by TomAFrench** [comment]
> comptime has been removed because it is no longer needed for accessing arrays.

It's still cheaper to perform comptime array accesses, no?

**Comment by jfecher** [comment]
> It's still cheaper to perform comptime array accesses, no?

It is cheaper, but this was determined by how many optimizations the ssa could do rather than whether it was classified as such by the type checker. So the runtime is unchanged but it may be more difficult for users to track whether something is comptime in complex cases.

**Comment by jfecher** [comment]
@phated this should not be a breaking change, `comptime` is still in the language but is ignored and issues a warning when used now.

**Comment by phated** [comment]
The PR title should indicate that then. I'll update.

**Comment by jfecher** [comment]
@phated it's more than just warning though since `comptime` is removed from the compiler internally. Perhaps `Deprecate comptime`? It is removed from the language itself, it is just that we keep parser support for it to issue deprecation warnings

**Comment by phated** [comment]
That's a tough one, I think I'd call it breaking then since it changes programs. `feat!: Remove `comptime` and issue removal warning upon usage`

**Comment by jfecher** [comment]
This shouldn't break any existing programs and the circuits generated should be unchanged as well since `comptime` in the type system never affected codegen

**Comment by Savio-Sou** [comment]
cc @critesjosh @signorecello in case if you have any thoughts from external devs' perspectives on the PR.

File: crates/noirc_evaluator/src/ssa/ir/dfg.rs [hunk 1]
```diff
@@ -408,11 +408,11 @@ impl DataFlowGraph {
     }
 
     pub(crate) fn get_location(&self, id: &InstructionId) -> Option<Location> {
-        self.locations.get(id).cloned()
+        self.locations.get(id).copied()
     }
 
     pub(crate) fn get_value_location(&self, id: &ValueId) -> Option<Location> {
-        match &self.values[*id] {
+        match &self.values[self.resolve(*id)] {
```
Comment (kevaundray) [inline crates/noirc_evaluator/src/ssa/ir/dfg.rs:415]: oof

File: crates/noirc_evaluator/src/ssa/opt/unrolling.rs [hunk 2]
```diff
@@ -547,7 +577,7 @@ mod tests {
         // }
         // The final block count is not 1 because unrolling creates some unnecessary jmps.
         // If a simplify cfg pass is ran afterward, the expected block count will be 1.
-        let ssa = ssa.unroll_loops();
+        let ssa = ssa.unroll_loops().unwrap();
```
Comment (kevaundray) [inline crates/noirc_evaluator/src/ssa/opt/unrolling.rs:580]: nit: can you use expect?

File: crates/noirc_frontend/src/hir_def/types.rs [hunk 3]
```diff
@@ -405,225 +402,9 @@ impl TypeBinding {
...
-
-            (CompTime::Maybe(_, binding), other) | (other, CompTime::Maybe(_, binding))
-                if binding.borrow().is_some() =>
-            {
-                let binding = &*binding.borrow();
-                binding.as_ref().unwrap().and(other, span)
-            }
-
-            (CompTime::Maybe(id1, _), CompTime::Maybe(id2, _)) if id1 == id2 => self.clone(),
-
```
Comment (kevaundray) [inline crates/noirc_frontend/src/hir_def/types.rs:592]: Sea of red :)

File: crates/noirc_frontend/src/hir_def/types.rs [hunk 4]
```diff
@@ -939,91 +664,65 @@ impl Type {
...
                     },
                 }
             }
-            _ => Err(SpanKind::None),
+            _ => Err(UnificationError),
         }
     }
 
     /// Try to bind a PolymorphicInt variable to self, succeeding if self is an integer, field,
     /// other PolymorphicInt type, or type variable. If use_subtype is true, the CompTime fields
```
Comment (kevaundray) [inline crates/noirc_frontend/src/hir_def/types.rs:678]: This still mentions `CompTime`

**Review by kevaundray — DISMISSED** [review]
LGTM - some small nits

File: crates/noirc_frontend/src/hir_def/types.rs (hunk 4 above)
Comment (jfecher) [reply crates/noirc_frontend/src/hir_def/types.rs]: I went through and removed all comment references and a few errors that aren't produced anymore

File: crates/noirc_evaluator/src/ssa/opt/unrolling.rs (hunk 2 above)
Comment (jfecher) [reply crates/noirc_evaluator/src/ssa/opt/unrolling.rs]: Updated

File: crates/noirc_evaluator/src/ssa/ir/dfg.rs (hunk 1 above)
Comment (jfecher) [reply crates/noirc_evaluator/src/ssa/ir/dfg.rs:415]: I'm not sure if this is needed but I added it just in case.

**Review by kevaundray — APPROVED** [review]

**Pull request merged**